src/services/mod.rs

```diff
@@ -20,3 +22,12 @@ pub fn offset_page_meta(meta: &PageMetadata, offset: i64) -> String {
     }
     .to_url_string()
 }
+
+pub async fn backend_url(
+    chain_id: &str,
+    info_provider: impl InfoProvider,
+    path_query: impl Fn() -> String,
```

Comment:
For me it feels weird to pass a function for this. This would be a very nice use case to put this into a macro

Reply:
I agree with you but after reading a bit about macros, there are certain things that make me think that this might not be a good use case for them. As I understand it, the nice thing to have here would be `core_backend_url!(chain_id, infor_provider, varargs)` , right? Basically having a variadic function.

`macros` don't have type checks and in general wouldn't prevent us from passing the arguments in whichever order, unlike regular functions do. Which makes me think, how do I even know which variable in the macro is the `info_provider` ? I found this example: https://www.reddit.com/r/learnrust/comments/ipbtge/how_to_generate_async_methods_with_macro/

Where I could technically try passing the `InfoProvider` trait as a first argument and even have an async context. I am not sure how this would play out, but yeah, I've read a bit as well that macros would slow down compilation time, even though right now it is not an issue.

Maybe the callback isn't the nices solution. Maybe I could pass on the path parameters as a `&[&str]` , if you prefer that in terms of style. I personally prefer the callback.